### ISSUE TITLE
Proposal: Type-safe and thread-safe Delegate helper methods

### ISSUE THREAD

**Issue by HaloFour** [dotnet/runtime#49619]
## Background and Motivation

The delegate API to combine or remove delegate instances from a multicast delegate hasn't really changed since 1.x, before generics.  Even after generics were added to the runtime new APIs weren't added likely because C# didn't support `System.Delegate` as a generic type constraint.  Now that C# does allow `System.Delegate` as a generic type constraint I think it's worthwhile considering type-safe versions of these APIs.

Furthermore, as delegates are immutable and the instance must be replaced outright it can create a race condition if you have more than one thread that can be subscribing or unsubscribing simultaneously.  Languages like C# and VB.NET will deal with that for the developer for auto-implemented events, but in those cases where a developer either needs an explicitly implemented event or is managing delegate instances outside of events they need to implement the same kind of logic to avoid this race condition.  Again, now that C# does allow `System.Delegate` as a generic type constraint I think it's worthwhile adding new APIs to modify a `ref` of a delegate instance that can do so in a thread-safe manner.

## Proposed API

```cs
public class Delegate {
    public static T? Combine<T>(T? a, T? b) where T : System.Delegate {
        return (T?) Delegate.Combine(a, b);
    }

    public static T? Remove<T>(T? a, T? b) where T : System.Delegate {
        return (T?) Delegate.Remove(a, b);
    }

    public static T? Combine<T>(ref T? field, T? value) where T : System.Delegate {
        T? original = field;
        while (true) {
            T? current = original;
            T? proposed = Delegate.Combine<T>(current, value); 
            T? previous = Interlocked.CompareExchange(ref field, proposed, current);
            if (previous == original) {
                return proposed;
            }
            original = previous;
        }
    }

    public static T? Remove<T>(ref T? field, T? value) where T : System.Delegate {
        T? original = field;
        while (true) {
            T? current = original;
            T? proposed = Delegate.Remove<T>(current, value); 
            T? previous = Interlocked.CompareExchange(ref field, proposed, current);
            if (previous == original) {
                return proposed;
            }
            original = previous;
        }
    }
}
```

## Usage Examples

```cs
public class MyClass {
    private EventHandler _myEvent;

    public event EventHandler MyEvent {
        add {
            // do something here
            Delegate.Combine<EventHandler>(ref _myEvent, value);
        }
        remove {
            // do something here
            Delegate.Remove<EventHandler>(ref _myEvent, value);
        }
    }
}
```

## Alternative Designs

The generic versions of `Combine` and `Remove` aren't required to facilitate the versions that can modify a `ref` delegate, but I do believe that they are useful type-safe additions in their own right.

**Comment by GrabYourPitchforks** [comment]
In your usage example, wouldn't it be easier to define a private event field and to have your public handlers wrap the private event handlers? This doesn't require any new API.

```cs
public class MyClass {
    private event EventHandler _myEvent;

    public event EventHandler MyEvent {
        add {
            // do something here
            _myEvent += value;
        }
        remove {
            // do something here
            _myEvent -= value;
        }
    }
}
```

**Comment by HaloFour** [comment]
@GrabYourPitchforks 

Yes, but that still relies on the compiler emitting a properly thread-safe implementation of those add/remove accessors.  I doubt that's something most C# devs would think to do in that situation.  If you're not working with events you're still on the hook figuring out thread-safe delegate manipulation.

/cc @CyrusNajmabadi

**Comment by GrabYourPitchforks** [comment]
Hmm, interesting point. Maybe this would be of interest to people writing code generators? I wonder if there are other "let the compiler work magic" patterns that it would make sense to promote to full-featured APIs.

Offhandedly, the pattern I mentioned previously wouldn't work for an array of events (think attempting to manipulate a single element of an `EventHandler[]` array). In that case you'd need a ref-based API as you had suggested. That's probably a weird edge case though.

**Comment by jkotas** [comment]
This is partial duplicate of https://github.com/dotnet/runtime/issues/4556 . The compiler magic does not actually work in cases today.
